Fix issue  re: DDP and create_graph=True

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63831

Closes https://github.com/pytorch/pytorch/issues/63812

`at::mul_out` is not supported when `grad` itself requires grad, which is useful for computing higher order derivatives.

In this case, fall back to a mul + copy instead of mul_out.

Note that DDP doesn't really work well with higher order grads. Will add appropriate warning messaging.

Differential Revision: [D30505573](https://our.internmc.facebook.com/intern/diff/D30505573/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30505573/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

rohan-varma committed:
Fix issue  re: DDP and create_graph=True

Closes https://github.com/pytorch/pytorch/issues/63812

`at::mul_out` is not supported when `grad` itself requires grad, which is useful for computing higher order derivatives.

In this case, fall back to a mul + copy instead of mul_out.

Differential Revision: [D30505573](https://our.internmc.facebook.com/intern/diff/D30505573/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30505573/)!

[ghstack-poisoned]

rohan-varma commented:
cc @mcarilli

rohan-varma committed:
Update on "Fix issue  re: DDP and create_graph=True"


Closes https://github.com/pytorch/pytorch/issues/63812

`at::mul_out` is not supported when `grad` itself requires grad, which is useful for computing higher order derivatives.

In this case, fall back to a mul + copy instead of mul_out.

Differential Revision: [D30505573](https://our.internmc.facebook.com/intern/diff/D30505573/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30505573/)!

[ghstack-poisoned]

zhaojuanmao commented:
I thought DDP does not support grad's gradients, e.g., the grad's gradients are not synced in current DDP. possibly we should be careful to support this new feature

rohan-varma commented:
> I thought DDP does not support grad's gradients, e.g., the grad's gradients are not synced in current DDP. possibly we should be careful to support this new feature

Yes, grad of grads will not be synced and I don't think that is a supported feature of DDP. Although, it seems like we do not intend to crash here so possibly we should fix that. Maybe we can clarify in the documentation that grads resulting from higher-order differentiation won't be synced.

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -377,9 +377,16 @@ void Reducer::mark_variable_ready_dense(size_t variable_index) {
+            // all the input parameters.
+            at::mul_out(bucket_view, grad, wrapped);
+          } else {
+            // If DDP is running with create_graph=True, gradients require_grad
```
I see, you mean codes will crash here if grad.requires_grad()? if so, makes sense to fix that here, but would you please also add a warning here saying it is not supported by DDP to sync up its gradients

rohan-varma replied:
sure, will add it

rohan-varma committed:
Update on "Fix issue  re: DDP and create_graph=True"


Closes https://github.com/pytorch/pytorch/issues/63812

`at::mul_out` is not supported when `grad` itself requires grad, which is useful for computing higher order derivatives.

In this case, fall back to a mul + copy instead of mul_out.

Note that DDP doesn't really work well with higher order grads. Will add appropriate warning messaging.

Differential Revision: [D30505573](https://our.internmc.facebook.com/intern/diff/D30505573/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30505573/)!

[ghstack-poisoned]
